File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/TensorArray.scala
/*
 * Copyright 2016 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.intel.analytics.bigdl.nn.ops

import java.util.concurrent.ConcurrentHashMap

import com.intel.analytics.bigdl.tensor.Tensor
import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
import com.intel.analytics.bigdl.utils.{RandomGenerator, T, Table}

import scala.collection.mutable.ArrayBuffer
import scala.reflect.ClassTag

private[nn] class TensorArrayInstance[D: ClassTag](
  val size: Int,
  val shape: Option[Seq[Int]] = None,
  val dynamicSize: Boolean = false,
  val clearAfterRead: Boolean = true,
  val identicalElementShapes: Boolean = false,
  val tensorArrayName: String = null)(implicit ev: TensorNumeric[D]) {

  private val tensors = new ArrayBuffer[Tensor[D]](size)

  def apply(index: Int): Tensor[D] = tensors(index)

  def update(index: Int, tensor: Tensor[D]): Unit = {
    tensors(index) = tensor
  }
}

class TensorArray[T: ClassTag, D: ClassTag](
  shape: Option[Seq[Int]] = None,
  dynamicSize: Boolean = false,
  clearAfterRead: Boolean = true,
  identicalElementShapes: Boolean = false,
  tensorArrayName: String = null
)(implicit ev: TensorNumeric[T], ev2: TensorNumeric[D])
  extends Operation[Tensor[Int], Table, T]{

  override def updateOutput(input: Tensor[Int]): Table = {
    val handle = if (tensorArrayName == null) {
      RandomGenerator.RNG.random().toString
    } else {
      tensorArrayName + RandomGenerator.RNG.random()
    }

    TensorArray.tensorArrayInstances.put(handle,
      new TensorArrayInstance[D](input.value(), shape, dynamicSize, clearAfterRead,
        identicalElementShapes, handle))
    output = T(
      Tensor.scalar[String](handle),
      Tensor.scalar[Float](0.0f)
    )
    output
  }
}

object TensorArray {
  private[nn] val tensorArrayInstances = new ConcurrentHashMap[String, TensorArrayInstance[_]]()

  def apply[T: ClassTag, D: ClassTag](
    shape: Option[Seq[Int]] = None,
    dynamicSize: Boolean = false,
    clearAfterRead: Boolean = true,
    identicalElementShapes: Boolean = false,
    tensorArrayName: String = null
  )(implicit ev: TensorNumeric[T], ev2: TensorNumeric[D]): TensorArray[T, D] =
    new TensorArray(shape, dynamicSize, clearAfterRead, identicalElementShapes, tensorArrayName)
}

class TensorArrayGrad[T: ClassTag, D: ClassTag](
  source: String
)(implicit ev: TensorNumeric[T], ev2: TensorNumeric[D])
  extends Operation[Table, Table, T]{
  override def updateOutput(input: Table): Table = {
    val handle = input[Tensor[String]](1)
    val flowIn = input[Tensor[Float]](2)

    val originalTensorArray = TensorArray.tensorArrayInstances.get(handle.value())
    TensorArray.tensorArrayInstances.put(handle.value() + source,
      new TensorArrayInstance[D](originalTensorArray.size,
        originalTensorArray.shape,
        originalTensorArray.dynamicSize,
        originalTensorArray.clearAfterRead,
        originalTensorArray.identicalElementShapes,
        handle.value() + source))
    output = T(
      Tensor.scalar[String](handle.value() + source),
      Tensor.scalar[Float](0.0f)
    )
    output
  }
}

object TensorArrayGrad {
  def apply[T: ClassTag, D: ClassTag](source: String)(
    implicit ev: TensorNumeric[T], ev2: TensorNumeric[D]): TensorArrayGrad[T, D] =
    new TensorArrayGrad(source)
}

class TensorArrayWrite[T: ClassTag, D: ClassTag]()(
  implicit ev: TensorNumeric[T], ev2: TensorNumeric[D])
  extends Operation[Table, Tensor[Float], T]{

  override def updateOutput(input: Table): Tensor[Float] = {
    val handle = input[Tensor[String]](1)
    val index = input[Tensor[Int]](2)
    val value = input[Tensor[D]](3)
    val flowIn = input[Tensor[Float]](4)

    val tensorArray = TensorArray.tensorArrayInstances.get(handle.value())
      .asInstanceOf[TensorArrayInstance[D]]
    tensorArray(index.value()) = value
    output = Tensor.scalar[Float](0.0f)
    output
  }
}

object TensorArrayWrite {
  def apply[T: ClassTag, D: ClassTag]()(
    implicit ev: TensorNumeric[T], ev2: TensorNumeric[D]): TensorArrayWrite[T, D] =
    new TensorArrayWrite()
}

class TensorArrayRead[T: ClassTag, D: ClassTag]()(
  implicit ev: TensorNumeric[T], ev2: TensorNumeric[D])
  extends Operation[Table, Tensor[D], T]{

  override def updateOutput(input: Table): Tensor[Float] = {
    val handle = input[Tensor[String]](1)
    val index = input[Tensor[Int]](2)
    val flowIn = input[Tensor[Float]](4)

    val tensorArray = TensorArray.tensorArrayInstances.get(handle.value())
      .asInstanceOf[TensorArrayInstance[D]]
    output = tensorArray(index.value())
    output
  }
}

object TensorArrayRead {
  def apply[T: ClassTag, D: ClassTag]()(
    implicit ev: TensorNumeric[T], ev2: TensorNumeric[D]): TensorArrayRead[T, D] =
    new TensorArrayRead()
}

File: spark/dl/src/test/resources/tf/models/tensor_array.py
#
# Copyright 2016 The BigDL Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import tensorflow as tf
import numpy as np

from sys import argv

from util import run_model

def main():
    tf.set_random_seed(10)
    with tf.Session() as sess:
        inputs = tf.Variable(tf.random_uniform((20, 30, 32)), name = 'input')
        inputs = tf.identity(inputs, "input_node")

        input1, input2, input3, input4 = tf.split(inputs, 4, 0)
        # scatter and gather
        tensor_array = tf.TensorArray(tf.float32, 128)
        tensor_array = tensor_array.scatter([1, 2, 5, 4, 3], input1)
        tensor_array.gather([1, 2, 5, 4, 3], name='scatter_and_gather')

        # split and concat
        tensor_array = tf.TensorArray(tf.float32, 2)
        tensor_array = tensor_array.split(input2, [2, 3])
        tf.identity(tensor_array.concat(), name='split_and_concat')

        # write and read
        tensor_array = tf.TensorArray(tf.float32, 5)
        tensor_array = tensor_array.identity()
        tensor_array = tensor_array.write(1, input3)
        tf.cast(tensor_array.size(), tf.float32, name='size1')
        tensor_array.read(1, name='write_and_read')
        tf.cast(tensor_array.size(), tf.float32, name='size2')

        # unstack and stack
        tensor_array = tf.TensorArray(tf.float32, 5)
        tensor_array = tensor_array.unstack(input4)
        tf.identity(tensor_array.stack(), name='unstack_and_stack')

        net_outputs = map(lambda x: tf.get_default_graph().get_tensor_by_name(x), argv[2].split(','))
        run_model(net_outputs, argv[1], None, argv[3] == 'True')

if __name__ == "__main__":
    main()
